### ISSUE TITLE
Compass with phone in vertical position is erratic

### ISSUE THREAD

**Issue by hevanw** [cgeo/cgeo#5889]
##### Detailed steps causing the problem:

Start navigating with the compass while the phone is held horizontally. This works very well. Now move the phone to a vertical position. The goal is to keep the phone in my shirt pocket while having spoken instructions on.
##### Actual behavior after performing these steps:

The direction (x o'clock) is often totally off (EDIT: actually it is probably off by 180 degrees).
When trying to reproduce, it looks like the compass behaves fine when the phone is held anywhere between -90 degrees (phone upside down, screen pointing forward) across 0 degrees (phone horizontally, screen pointing upward) to 90 degrees (phone upright, screen pointing towards yourself).
However, as soon as you move it past the 90 degree position (i.e. slightly leaning it toward you) the compass jumps to a different direction and literally has lost north.
This makes it impossible to use the compass directions while having the phone in a shirt pocket as it will often report the wrong direction.

Other compass apps on my phone behave correctly.
##### Expected behavior after performing these steps:

Compass should point to the right direction not only when held at 90 degrees or less, but also when held at more than 90 degrees.
##### Version of c:geo used:

2016.06.09
##### Is the problem reproducible for you?

Yes
##### System information:

--- System information ---
Device: ONEPLUS A3003 (OnePlus3, OnePlus)
Android version: 6.0.1
Android build: ONEPLUS A3003_16_160725
Cgeo version: 2016.06.09
Google Play services: enabled
Low power mode: inactive
Compass capabilities: yes
Rotation vector sensor: present
Orientation sensor: present
Magnetometer & Accelerometer sensor: present
Direction sensor used: rotation vector
Hide own/found: false
Map strategy: auto
HW acceleration: enabled (default state)
System language: nl_BE
Geocaching sites enabled:
- geocaching.com: logged in (Ingelogd) / BASIC
  Installed cgeo plugins:  none

--- End of system information ---

**Comment by Lineflyer** [comment]
Well, how could a magnetic compass in a smartphone work any better than a classic compass. Both will only work in horizontal position.

For your usecase it would be a suitable solution to use the GPS walking direction instead of the magnetic compass.
This can be easily done by switching off the magnetic compass on the compass screen before placing it into your pocket.

**Comment by hevanw** [comment]
Unlike a real compass, a smartphone has magnetometers in all 3 directions, so a compass app can perfectly work in positions other than horizontally. If you want to learn more, just check the free 'Physics Toolbox' app and play around a bit with the included Magnetometer.
Like I said, other compass apps have no problem when holding the phone in a vertical position.

**Comment by samueltardieu** [comment]
@Lineflyer: @hevanw is right, a MEMS compass has nothing to do with a physical one.

@hevanw: putting the phone vertical in your shirt pocket and expect the compass to work is difficult. As your phone will tilt forward and back, the calculated direction facing the phone will switch 180° at once. You can test with, for example, GPS status. If the upper side of your phone is slightly farther from you than the bottom side, it will indicate the "right" direction. However, if you tilt the phone a bit and make the upper side closer to you while the bottom side gets away from you (as if you were lying down while looking at your phone), the direction will change. This may be the origin of your problem.

Note that we don't do most of the direction computation ourselves, we use what Android gives us.

**Comment by hevanw** [comment]
Thanks for the explanation. I tried out several other compass apps and it appears there are different implementations/behaviors. It basically all comes down to whether the compass shows (a) your own orientation, or (b) the orientation of the phone. So, like you explained in case (b), as soon as you tilt the phone beyond 90 degrees, the phone is now pointing at your own 6 o'clock and the compass changes 180 degrees. There are however apps (ex: https://play.google.com/store/apps/details?id=com.apksoftware.compass ) that don't change and when moving the phone above your head with the screen pointing downwards, will still show the compass relative to yourself rather than to the phone. You could definitely argue that the compass is wrong then, since e.g. if you are facing N, the compass will show the top of the screen being N, but since the phone is upside down, the virtual compass is in fact pointing to the S.

In the context of c:geo, I think implementation (a) is more useful, especially in combination with the spoken instructions. But clearly it's now more of a feature request than a bug. It could even get better if c:geo could detect that I have my phone upside-down in my pocket (so the speaker is facing up) and still give the right instructions.

**Comment by samueltardieu** [comment]
Speaker facing up is the easy one :-) What about the screen? In your shirt pocket, the screen will either point towards your destination, or to your body (more likely as to protect the screen). In your pants pocket, it will globally either be pointing to your left or your right. That's why in this mode, where you do not look at the screen, it is best to enable GPS-based orientation which considers your move globally while ignoring the orientation of the phone itself.

**Comment by hevanw** [comment]
Yes, those are things that I did think about already :). I would always have to put my phone in my pocket with the screen facing my body (unless there would be an option to reverse that). You are right about pants pockets that then the phone probably won't even be perpendicular to my facing direction anyway.

Correct me if I'm wrong, but GPS based instructions are only usable if there are actual streets/paths on the map, which may not be the case if you go geocaching in more rough areas. Anyway, the compass based spoken instructions are quite neat, but they are rather pointless if at the same time you are supposed to keep your phone horizontally in your hand. For now, I'll see if I can find something that kinda forces the phone in an angled position in my shirt/jack pocket so it stays away from the vertical position :).

**Comment by samueltardieu** [comment]
> Correct me if I'm wrong, but GPS based instructions are only usable if there are actual streets/paths on the map

You're wrong. It works as long as you receive the GPS signal, which is needed anyway to compute the distance and the direction of the cache. No need for a street, for a map, or anything else. You might as well be offline without any network connection.

**Comment by Bananeweizen** [comment]
Use the GPS signal with the compass (i.e. NOT only the magnetic sensor), and walk with normal walking speed. The spoken instructions for navigating to the cache will be correct, independent of the positioning of your device. That is because the GPS receiver calculates the direction of travel by sampling your position over time. As long as you are fast enough (and to my experience normal walking is fast enough, at least on my phone), the direction will be indicated precise enough to give a direction like "2 o clock" and the cache actually being in the range of 1 to 3 o'clock. I've tested that extensively when I implemented the talking compass.
